utils/python-rpc/framework/daemon.py

```diff
@@ -32,10 +32,11 @@
 
 class Daemon(object):
 
-    def __init__(self, protocol='http', host='127.0.0.1', port=0, idx=0):
+    def __init__(self, protocol='http', host='127.0.0.1', port=0, idx=0, restricted_rpc = False):
+        base = 18480 if restricted_rpc else 18180
         self.host = host
         self.port = port
-        self.rpc = JSONRPC('{protocol}://{host}:{port}'.format(protocol=protocol, host=host, port=port if port else 18180+idx))
+        self.rpc = JSONRPC('{protocol}://{host}:{port}'.format(protocol=protocol, host=host, port=port if port else base+idx))
```

Comment:
That said, I don't have a strong dislike, so if you'd rather do it this way it can stay (and I'll move unrestricted to 2 later)

Reply:
Creating a new daemon index, and 3 wallets (for the transfer tests now using daemon 2).

Reply:
Why do you need 3 more wallets now ? That makes me think maybe I missed the reason why you did it the way you did...

Reply:
Sorry, repost to proper thread:

The transfer tests use three seeded wallets. I added checks to probe a restricted rpc to verify that the transactions in the local/stem phase don't show up.

Reply:
I don't get why you'd need to dupe all the wallets. "Restricted" is a daemon thing, not a wallet thing.

Reply:
The transfer test is now verifying that the wallet initiated send is "hidden" from restricted RPC calls. The wallets are hard-coded to a specific daemon at startup - the `Wallet()` constructor does not launch a new wallet process. So the wallets are initiating a send, and then calls are made to the unrestricted and restricted RPC ports to verify the different states.

Alternatively, the new daemon (index 2) could be launched that connects via p2p to the original daemon (index 0). There are two difficulties with this mode. The daemon-0 must connect to daemon-2 (for outgoing Dandelion++ stem), and daemon-2 p2p must be listening when daemon-0 launches (daemon-0 tries every second but the test is a bit more racy unless a wait for daemon-2 is added). This can be done by launching in reverse order with some sleeps and port probing. The second issue that if daemon-0 rolls "fluff", then the forwarding to the next daemon is randomly delayed by 1-4 seconds 95% of the time (there are no _actual_ bounds since its purely exponential). Additional sleeps in the transfer test could be added to handle this and be mostly effective, but they would need to be more than 4 second sleeps.

Adding new wallets seemed easier. Doing the alternative would (kind-of) functionally test Dandelion++. Realistically would need more daemons to fully test it though.

Reply:
Fair enough.